CARP/CBASS Uniform Weight Bug Fix

jjn13 opened the pull request:
This pulls request addresses:

https://github.com/DataSlingers/clustRviz/pull/15

for the develop branch.

Both CARPVIZ and CBASSVIZ are changed to allow one final iteration in the case that try_iter exceeds the ti threshold. This ensures the proper sparsity pattern is recorded; the link above contains a more detailed explanation.

Tests for CARP and CBASS are also added to verify that the first and returned final sparsity patterns are correct (no sparsity and full sparsity, respectively).

michaelweylandt commented:
Great! I'll take a closer look at this tomorrow - I want to spend some time convincing myself that it provably i) stops with a fully connected graph; ii) can't get into an infinite loop; iii) always returns a dendrogram.

michaelweylandt commented on src/cbass_viz.cpp:
```diff
@@ -172,6 +172,10 @@ Rcpp::List CBASS_VIZcpp(const Eigen::VectorXd& x,
     while(rep_iter){
+      // Safety check - only so many iterations of the inner loop before we move on
+      if(try_iter > ti){
+        rep_iter = false;
```
Using a `break` statement here may be cleaner than checking for `rep_iter` in every branch below.

michaelweylandt commented on src/carp_viz.cpp:
```diff
@@ -115,6 +115,10 @@ Rcpp::List CARP_VIZcpp(const Eigen::VectorXd& x,
     while(rep_iter){
+      // Safety check - only so many iterations of inner loop before we move on
+      if(try_iter > ti){
+        rep_iter = false;
```
Using a break statement here may be cleaner than checking for rep_iter in every branch below.

michaelweylandt commented on tests/testthat/test_carp_clustering.R:
```diff
@@ -63,3 +63,65 @@ test_that("Supplying percent and k gives error", {
+  carp.fit <- CARP(presidential_speech)
+  expect_equal(
+    sum(carp.fit$carp.sol.path$v.zero.inds[,ncol(carp.fit$carp.sol.path$v.zero.inds)]),
+    nrow(carp.fit$carp.sol.path$v.zero.inds)
```
Would it be better to test something like `all(carp_fit$... == 1)`?

michaelweylandt commented on tests/testthat/test_carp_clustering.R:
```diff
@@ -63,3 +63,65 @@ test_that("Supplying percent and k gives error", {
+  carp.fit <- CARP(presidential_speech,alg.type='carpviz')
+  expect_equal(
+    sum(carp.fit$carp.sol.path$v.zero.inds[,ncol(carp.fit$carp.sol.path$v.zero.inds)]),
+    nrow(carp.fit$carp.sol.path$v.zero.inds)
```
Isn't this the same test as above? (I.e., isn't `carpviz` our default?)

michaelweylandt commented on tests/testthat/test_carp_clustering.R:
```diff
@@ -63,3 +63,65 @@ test_that("Supplying percent and k gives error", {
+
+
+
+test_that("CARP-returned sparsity pattern begins with no fusions (uniform weights)",{
```
Is this the right thing to check? What happens if, *e.g.*, we have two data points which are equal in all features (e.g., if GCleveland gets two rows instead of one)? Shouldn't they be fused at $\lambda = 0$ anyways?

jjn13 replied:
This is true. Perhaps we exclude these columns for testing in the case this happens? If so, I'll modify test to check for this.

michaelweylandt replied:
In the context of this test, it's no big deal. Just want to determine if this is something we're sensitive to generally.

michaelweylandt commented on tests/testthat/test_carp_clustering.R:
```diff
@@ -63,3 +63,65 @@ test_that("Supplying percent and k gives error", {
+  carp.fit <- CARP(presidential_speech,alg.type = 'carpviz')
+  expect_equal(
+    sum(carp.fit$carp.sol.path$v.zero.inds[,1]),
+    0
```
Perhaps `all(... == 0)` instead?

michaelweylandt commented on tests/testthat/test_carp_clustering.R:
```diff
@@ -63,3 +63,65 @@ test_that("Supplying percent and k gives error", {
+})
+
+test_that("CARPVIZ-returned sparsity pattern begins with no fusions",{
+  carp.fit <- CARP(presidential_speech,alg.type = 'carpviz')
```
Space after comma (here and several other places)

michaelweylandt commented on tests/testthat/test_carp_clustering.R:
```diff
@@ -63,3 +63,65 @@ test_that("Supplying percent and k gives error", {
+    0
+  )
+})
+test_that("CARPVIZ-returned sparsity pattern ends with full fusions (uniform weights)",{
```
As above - I think some of these are redundant since `carpviz` is the default algorithm. Did you mean to set `carp` and `t` explicitly earlier?

michaelweylandt commented on tests/testthat/test_cbass_clustering.R:
```diff
@@ -141,3 +141,149 @@ test_that("CBASS clustering cluster.mean.matrix respects variable/column labels"
   expect_equal(colnames(presidential_speech),
                colnames(clustering(cbass_fit, percent = .3)$cluster.mean.matrix))
 })
+
```
All comments in `CARP` tests apply here also.

michaelweylandt commented:
Copying from #15 for ease of access: 

@jjn13 said: 

---

For the backtracking search, the `ti` parameter defines the maximum number of backtracking   attempts.

On each attempt (`try_iter`), we reduce the amount of regularization to see which fusion happened first (here assuming that multiple fusions occurred, since this is the reason we are backtracking in the first place).

If we don't succeed, this places us in the second condition of the `if/else` clauses and
`vZeroIndsnew` is set back to `vZeroIndsold` (designed to make us ready for the next try).

However if we've already exceeded `ti`, we don't get to try again, and `vZeroIndsnew` is not included later in `vZeroIndsPath` for the current iteration.

For the uniform weight example this occurred on the last iteration leaving out the final, fully fused sparsity pattern, breaking dendrogam code downstream.

Moving the `ti` check up front and including the `rep_iter` condition in the `if/else`'s makes sure that we take one final pass, knowing we won't get to retry again.

---

A few higher-level questions: 

- What level of `gamma` do we want the final "insurance" iteration to be at? 
- In the second branch[1], we handle the first iteration separately, but we don't in the third. Why is that? 
- It looks like a comment got clipped at [2]; could you fix that up? 
- Any intuition on why this happened in the later iteration of the uniform case? That seems pretty non-pathological.
- The logic is a bit tricky to follow because we have `old` and `new` versions of our variables. Would a third name help (fixing `old` at the previous iteration) - maybe `working`?

[1] https://github.com/DataSlingers/clustRviz/pull/39/commits/2634dde1ea899f18c3baf7b69952346a98498de0#diff-78a7eab5b2a5542faf4de988926219e1R256

[2] https://github.com/DataSlingers/clustRviz/pull/39/commits/2634dde1ea899f18c3baf7b69952346a98498de0#diff-78a7eab5b2a5542faf4de988926219e1R263

michaelweylandt commented:
This needs to be rebased against `develop` to avoid conflicts. 

@jjn13 While you're rebasing, could you also clean up the history (apply the last 4 commits to the original commits that added that code) so that things are a bit cleaner?

michaelweylandt commented:
@jjn13 Any updates on this PR?

jjn13 commented:
I rebased this against develop and consolidated the previous commits.

A lot of the tests previously in test_{carp,cbass}_clustering.R are now in the accessors tests, so I removed them. The test files now just look at the carp/cbass sparsity patterns, and not the return values of the get_cluster_labels, get_cluster_centroids.

The carp/cbass-viz try_iter correction remains the same.

michaelweylandt commented:
@jjn13 Any thoughts on these questions? (repeated from above)

- What level of gamma do we want the final "insurance" iteration to be at? 
- In the second branch[1], we handle the first iteration separately, but we don't in the third. Why is that? 
- It looks like a comment got clipped at [2]; could you fix that up? 
- Any intuition on why this happened in the last iteration of the uniform case? That seems pretty non-pathological. 
- The logic is a bit tricky to follow because we have old and new versions of our variables. Would a third name help (fixing old at the previous iteration) - maybe working?  

[1] https://github.com/DataSlingers/clustRviz/commit/2634dde1ea899f18c3baf7b69952346a98498de0#diff-78a7eab5b2a5542faf4de988926219e1R256
[2] https://github.com/DataSlingers/clustRviz/commit/2634dde1ea899f18c3baf7b69952346a98498de0#diff-78a7eab5b2a5542faf4de988926219e1R263
